packages/gatsby/src/utils/worker/child/queries.ts

```diff
@@ -29,7 +35,39 @@ function getGraphqlRunner(): GraphQLRunner {
   return gqlRunner
 }
 
-export async function runQueries(queryIds: IGroupedQueryIds): Promise<void> {
+type ActionsToReplay = Array<
+  | IQueryStartAction
+  | IPageQueryRunAction
+  | IAddPendingPageDataWriteAction
+  | ICreatePageDependencyAction
+>
+
+export async function runQueries(
+  queryIds: IGroupedQueryIds
+): Promise<ActionsToReplay> {
+  const actionsToReplay: ActionsToReplay = []
+
+  const unsubscribe = store.subscribe(() => {
+    const action = store.getState().lastAction
+    if (
+      action.type === `QUERY_START` ||
+      action.type === `PAGE_QUERY_RUN` ||
+      action.type === `ADD_PENDING_PAGE_DATA_WRITE` ||
+      action.type === `CREATE_COMPONENT_DEPENDENCY`
+    ) {
+      actionsToReplay.push(action)
+    }
+  })
+
+  try {
+    await doRunQueries(queryIds)
+    return actionsToReplay
+  } finally {
+    unsubscribe()
+  }
```

Comment:
If I understand correctly this will discard any actions that happened during the batch execution if one query fails. I think we are fine with rerunning succesfull queries in subsequent builds as long as that's how interaction works? Are queries marked as dirty until main process replay those actions?

Reply:
I think if one query fails we fail the whole build now?

Reply:
> I think if one query fails we fail the whole build now?

Yeah, I just meant the subsequent build if cache was not cleared or invalidated (for example fixing something in markdown file that caused query to fail).

> Also, if queries are executed by the worker - they are already marked as dirty. And since we don't replay those actions - they stay dirty for the main process.

That was my question :) Wether queries remain marked as dirty in main process until we replay query related actions received from worker(s).